src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormattingPass.cs

```diff
@@ -110,13 +105,124 @@ private async Task<List<TextEdit>> FormatCSharpAsync(FormattingContext context,
+
+            // First, let's build an understanding of the desired C# indentation at the beginning and end of each source mapping.
+            var sourceMappingIndentations = new SortedDictionary<int, int>();
+            foreach (var mapping in context.CodeDocument.GetCSharpDocument().SourceMappings)
+            {
+                var startIndentation = CSharpFormatter.GetCSharpIndentation(context, mapping.GeneratedSpan.AbsoluteIndex, cancellationToken);
+                sourceMappingIndentations[mapping.OriginalSpan.AbsoluteIndex] = startIndentation;
+
+                var endIndentation = CSharpFormatter.GetCSharpIndentation(context, mapping.GeneratedSpan.AbsoluteIndex + mapping.GeneratedSpan.Length + 1, cancellationToken);
+                sourceMappingIndentations[mapping.OriginalSpan.AbsoluteIndex + mapping.OriginalSpan.Length + 1] = endIndentation;
+            }
+
+            var sourceMappingIndentationScopes = sourceMappingIndentations.Keys.ToArray();
+
+            // Now, let's combine the C# desired indentation with the Razor and HTML indentation for each line.
+            var newIndentations = new Dictionary<int, int>();
+            for (var i = range.Start.Line; i <= range.End.Line; i++)
+            {
+                var line = context.SourceText.Lines[i];
+                if (line.Span.Length == 0)
+                {
+                    // We don't want to indent empty lines.
+                    continue;
+                }
+
+                var lineStart = line.Start;
+                int csharpDesiredIndentation;
+                if (DocumentMappingService.TryMapToProjectedDocumentPosition(context.CodeDocument, lineStart, out _, out var projectedLineStart))
+                {
+                    // We were able to map this line to C# directly.
+                    csharpDesiredIndentation = CSharpFormatter.GetCSharpIndentation(context, projectedLineStart, cancellationToken);
+                }
+                else
+                {
+                    // Couldn't remap. This is probably a non-C# location.
+                    // Use SourceMapping indentations to locate the C# scope of this line.
+                    // E.g,
+                    // 
+                    // @if (true) {
+                    //   <div>
+                    //  |</div>
+                    // }
+                    //
+                    // We can't find a direct mapping at |, but we can infer its base indentation from the
+                    // indentation of the latest source mapping prior to this line.
+                    // We use binary search to find that spot.
+                    
+                    var index = Array.BinarySearch(sourceMappingIndentationScopes, lineStart);
+                    if (index < 0)
+                    {
+                        // Couldn't find the exact value. Find the index of the element to the right of the searched value.
+                        index = ~index;
```

Comment:
lol why negate -1 vs. just setting it?

Reply:
This is how the binary search API works. From the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.array.binarysearch?view=netcore-3.1),

> If the Array does not contain the specified value, the method returns a negative integer. You can apply the bitwise complement operator (~ in C#, Not in Visual Basic) to the negative result to produce an index. If this index is one greater than the upper bound of the array, there are no elements larger than value in the array. Otherwise, it is the index of the first element that is larger than value.

Or did I misunderstand your question?

Reply:
oh wow I didn't realize it had that. Nevermind